log/record.go

```diff
@@ -0,0 +1,183 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// Copyright 2022 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package log // import "go.opentelemetry.io/otel/log"
+
+import (
+	"errors"
+	"time"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+)
+
+var errUnsafeAddAttrs = errors.New("unsafely called AddAttrs on copy of Record made without using Record.Clone")
+
+// Record TODO: comment.
+// TODO: Add unit tests.
+type Record struct {
+	// TODO: comment.
+	Timestamp time.Time
+
+	// TODO: comment.
+	ObservedTimestamp time.Time
+
+	// TODO: comment.
+	Severity Severity
+
+	// TODO: comment.
+	SeverityText string
+
+	// TODO: comment.
+	Body string
+
+	// The fields below are for optimizing the implementation of
+	// Attributes and AddAttributes.
+
+	// Allocation optimization: an inline array sized to hold
+	// the majority of log calls (based on examination of open-source
+	// code). It holds the start of the list of attributes.
+	front [AttributesInlineCount]attribute.KeyValue
+
+	// The number of attributes in front.
+	nFront int
+
+	// The list of attributes except for those in front.
+	// Invariants:
+	//   - len(back) > 0 iff nFront == len(front)
+	//   - Unused array elements are zero. Used to detect mistakes.
+	back []attribute.KeyValue
+}
+
+// AttributesInlineCount TODO: comment that this should be used
+// by bridges to implement in order to reduce memory allocations.
+const AttributesInlineCount = 5
+
+// Severity TODO: comment.
+type Severity int
+
+// TODO: comment.
+const (
+	SeverityUndefined Severity = iota
+	SeverityTrace
+	SeverityTrace2
+	SeverityTrace3
+	SeverityTrace4
+	SeverityDebug
+	SeverityDebug2
+	SeverityDebug3
+	SeverityDebug4
+	SeverityInfo
+	SeverityInfo2
+	SeverityInfo3
+	SeverityInfo4
+	SeverityWarn
+	SeverityWarn2
+	SeverityWarn3
+	SeverityWarn4
+	SeverityError
+	SeverityError2
+	SeverityError3
+	SeverityError4
+	SeverityFatal
+	SeverityFatal2
+	SeverityFatal3
+	SeverityFatal4
+)
+
+// Attributes calls f on each [attribute.KeyValue] in the [Record].
+// Iteration stops if f returns false.
+func (r Record) Attributes(f func(attribute.KeyValue) bool) {
```

Comment:
1. This is done like that in order to minimize the heap allocations.
2. Data Model has to be implemented in SDK
3. I feel that "field" in the data model is more about "domain model" and it does not have to be implemented as a programming language field.

Reply:
> Adding a method to `LogRecord` will be a backwards compatible change. This is described in the design doc.

Right but how do you add an `Attribute` method to comply if you already have one?

Reply:
I understood the design choice when I made the comment. I'm still not convinced this is a good API from a users perspective.

Reply:
> I understood the design choice when I made the comment. I'm still not convinced this is a good API from a users perspective.

Any concrete things that you would like to change?

Side note (just for clarity). This is primary a bridge API and I think that in this case performance is more important than esthetics. So far I was not able to come up with better API while keeping the same amount of heap allocations when emitting a log record for 5 attributes.

> We found that over 95% of calls to logging methods pass five or fewer attributes

From https://go.dev/blog/slog

Related thread: https://github.com/open-telemetry/opentelemetry-go/pull/4725#discussion_r1400869566

Reply:
> Right but how do you add an `Attribute` method to comply if you already have one?

I think it would be good to rename the method to `WalkAttributes`. Safer and I think that more idiomatic (https://pkg.go.dev/path/filepath).

I do not expect such method would be added.

Reply:
What about adding "bound" methods to a `Logger` like `WithAttributes`, `WithName`, ...?

Reply:
> Any concrete things that you would like to change?

Unify how a user will set values  of a `Record`. Either all fields or all getters/setters.

Having a user do something like

```go
logger.Emit(ctx, Record{Body: "body"})
```

and then have them want to add an attribute so they do this:

```go
logger.Emit(ctx, Record{Body: "body", Attributes: []attribute{...}})
```

But the API of the `Record` in this case is confusing and requires refactoring.

Reply:
My preference is to have a `WithAttribute` on the `Logger`:

```go
type Logger interface {
    /*... */
    WithAttributes(/*...*/) Logger
}
```

And then just have an `Attributes` field in the `Record`.

```go
type Record struct {
   /* ... */
   Attributes []attribute.KeyValue
}
```

If a user of the bridge is performance sensitive they can use the `WithAttributes` method of the `Logger` to apply attributes to all emitted records. It will be up to the implementation to handle optimization, not the API.

This will be similar to the ideas talked about in the metric signal regarding a "bound" method: https://github.com/jmacd/opentelemetry-specification/blob/500965812d8942704ae2a60eb4ea69ca3df6337c/specification/metrics/api.md#bound-instrument-calling-convention

This would need to be added to the specification (i.e. allow an option for this method on the Logger similar to what was done in metrics).

Reply:
> If a user of the bridge is performance sensitive they can use the WithAttributes method of the Logger to apply attributes to all emitted records

1. The user can add different attributes for each emitted record. 
2. Adding even a single attribute for a record would make a heap allocation.

The current design makes it possible to have no allocations when the user adds  up to 5 attributes to the log record thanks to the array (instead of a slice) encapsulated in `Record`.

Reply:
I understand all this. I still feel like a user is smart enough to chose their own path here.

If they want a highly performant use of the bridge they can call `WithAttributes` where needed. If they don't, the fact that attributes can be passed as a field provides a concise and easy way to call `Emit` in a single line.

Reply:
What about going the same path as `slog` entirely and accepting `any` in `Emit`?

```go
type Logger interface {
    Emit(ctx context.Context, body string, args ...any)
}
```

Reply:
> What about going the same path as slog entirely and accepting any in Emit?

The Go escape analysis was not so clever in this scenario and it causes heap allocation.
This problem exists in `logr.LogSink` and you can see the allocs in `BenchmarkLogr`. I checked the code with `GODEBUG=allocfreetrace=1` and many of the allocations where done for the `args ...any` parameter.

Also for reference:

> For the most efficient log output, use [Logger.LogAttrs](https://pkg.go.dev/log/slog#Logger.LogAttrs). It is similar to [Logger.Log](https://pkg.go.dev/log/slog#Logger.Log) but accepts only Attrs, not alternating keys and values; this allows it, too, to avoid allocation.

From https://pkg.go.dev/log/slog#hdr-Attrs_and_Values

Reply:
> If they want a highly performant use of the bridge they can call `WithAttributes` where needed. If they don't, the fact that attributes can be passed as a field provides a concise and easy way to call Emit in a single line.

So the logger could have attributes array instead of the record? This is a smart idea worth exploring.

Reply:
At this point of time I think that neither `WithAttributes` nor `WithName` are really necessary.

Instead of `WithAttributes` the user can create a record with attributes and clone it.

Instead of `WithName` the user could create a new logger using `LoggerProvider`.

I will try to implement more methods in the log bridges to exercise that.

Reply:
Resolving.

I made a separate thread for `WithAttributes` to avoid some noise.